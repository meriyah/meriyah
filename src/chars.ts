/**
 * A list of character constants with much more human-readable names.
 */
export const enum Chars {

    Null = 0x00,

    Backspace      = 0x08,
    Tab            = 0x09,
    LineFeed       = 0x0A,
    VerticalTab    = 0x0B,
    FormFeed       = 0x0C,
    CarriageReturn = 0x0D,

    Space       = 0x20,
    Exclamation = 0x21,
    DoubleQuote = 0x22,
    Hash        = 0x23,
    Dollar      = 0x24,
    Percent     = 0x25,
    Ampersand   = 0x26,
    SingleQuote = 0x27,
    LeftParen   = 0x28,
    RightParen  = 0x29,
    Asterisk    = 0x2A,
    Plus        = 0x2B,
    Comma       = 0x2C,
    Hyphen      = 0x2D,
    Period      = 0x2E,
    Slash       = 0x2F,

    Zero         = 0x30,
    One          = 0x31,
    Two          = 0x32,
    Three        = 0x33,
    Four         = 0x34,
    Five         = 0x35,
    Six          = 0x36,
    Seven        = 0x37,
    Eight        = 0x38,
    Nine         = 0x39,
    Colon        = 0x3A,
    Semicolon    = 0x3B,
    LessThan     = 0x3C,
    EqualSign    = 0x3D,
    GreaterThan  = 0x3E,
    QuestionMark = 0x3F,

    UpperA = 0x41,
    UpperB = 0x42,
    UpperC = 0x43,
    UpperD = 0x44,
    UpperE = 0x45,
    UpperF = 0x46,
    UpperG = 0x47,
    UpperH = 0x48,
    UpperI = 0x49,
    UpperJ = 0x4A,
    UpperK = 0x4B,
    UpperL = 0x4C,
    UpperM = 0x4D,
    UpperN = 0x4E,
    UpperO = 0x4F,

    UpperP       = 0x50,
    UpperQ       = 0x51,
    UpperR       = 0x52,
    UpperS       = 0x53,
    UpperT       = 0x54,
    UpperU       = 0x55,
    UpperV       = 0x56,
    UpperW       = 0x57,
    UpperX       = 0x58,
    UpperY       = 0x59,
    UpperZ       = 0x5A,
    LeftBracket  = 0x5B,
    Backslash    = 0x5C,
    RightBracket = 0x5D,
    Caret        = 0x5E,
    Underscore   = 0x5F,

    Backtick = 0x60,
    LowerA   = 0x61,
    LowerB   = 0x62,
    LowerC   = 0x63,
    LowerD   = 0x64,
    LowerE   = 0x65,
    LowerF   = 0x66,
    LowerG   = 0x67,
    LowerH   = 0x68,
    LowerI   = 0x69,
    LowerJ   = 0x6A,
    LowerK   = 0x6B,
    LowerL   = 0x6C,
    LowerM   = 0x6D,
    LowerN   = 0x6E,
    LowerO   = 0x6F,

    LowerP      = 0x70,
    LowerQ      = 0x71,
    LowerR      = 0x72,
    LowerS      = 0x73,
    LowerT      = 0x74,
    LowerU      = 0x75,
    LowerV      = 0x76,
    LowerW      = 0x77,
    LowerX      = 0x78,
    LowerY      = 0x79,
    LowerZ      = 0x7A,
    LeftBrace   = 0x7B,
    VerticalBar = 0x7C,
    RightBrace  = 0x7D,
    Tilde       = 0x7E,
    Delete      = 0x7F,

    NextLine         = 0x85,
    NonBreakingSpace = 0xA0,

    Ogham = 0x1680,

    EnQuad             = 0x2000,
    EmQuad             = 0x2001,
    EnSpace            = 0x2002,
    EmSpace            = 0x2003,
    ThreePerEmSpace    = 0x2004,
    FourPerEmSpace     = 0x2005,
    SixPerEmSpace      = 0x2006,
    FigureSpace        = 0x2007,
    PunctuationSpace   = 0x2008,
    ThinSpace          = 0x2009,
    HairSpace          = 0x200A,
    ZeroWidthSpace     = 0x200B,
    ZeroWidthJoiner    = 0x200C,
    ZeroWidthNonJoiner = 0x200D,

    LineSeparator      = 0x2028,
    ParagraphSeparator = 0x2029,

    NarrowNoBreakSpace = 0x202F,
    MathematicalSpace  = 0x205F,
    IdeographicSpace   = 0x3000,

    ZeroWidthNoBreakSpace = 0xFEFF,

    ByteOrderMark = 0xFFEF,

    NonBMPMax = 0x10FFFF,

    LeadSurrogateMin = 0xD800,
    LeadSurrogateMax = 0xDBFF,
    TrailSurrogateMin = 0xDC00,
    TrailSurrogateMax = 0xDFFF,
    UTF16Max = 0xFFFF,
    NonBMPMin = 0x10000,
}
