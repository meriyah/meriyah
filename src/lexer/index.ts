export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
