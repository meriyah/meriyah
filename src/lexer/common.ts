import { Chars } from '../chars';
import { Flags } from '../common';
import { type Parser } from '../parser/parser';
import { Token } from '../token';

export const enum LexerState {
  None = 0,
  NewLine = 1 << 0,
  LastIsCR = 1 << 2,
}

export const enum NumberKind {
  ImplicitOctal = 1 << 0,
  Binary = 1 << 1,
  Octal = 1 << 2,
  Hex = 1 << 3,
  Decimal = 1 << 4,
  NonOctalDecimal = 1 << 5,
  Float = 1 << 6,
  ValidBigIntKind = 1 << 7,
  DecimalNumberKind = Decimal | NonOctalDecimal,
}

/**
 * Advances this lexer's current index.
 *
 * @param parser The parser instance
 */
export function advanceChar(parser: Parser): number {
  parser.column++;
  return (parser.currentChar = parser.source.charCodeAt(++parser.index));
}

/**
 * Consumes possible surrogate pair, return the merged code point or 0
 *
 * @param parser The parser instance
 */
export function consumePossibleSurrogatePair(parser: Parser): number {
  // See: https://tc39.github.io/ecma262/#sec-ecmascript-language-types-string-type
  const hi = parser.currentChar;
  // Not surrogate
  if ((hi & 0xfc00) !== Chars.LeadSurrogateMin) return Chars.Null;
  const lo = parser.source.charCodeAt(parser.index + 1);
  // Unpaired surrogate
  if ((lo & 0xfc00) !== Chars.TrailSurrogateMin) return Chars.Null;
  // Merged code point
  return Chars.NonBMPMin + ((hi & 0x3ff) << 10) + (lo & 0x3ff);
}

/**
 * Use to consume a line feed instead of `scanNewLine`.
 */
export function consumeLineFeed(parser: Parser, state: LexerState): void {
  parser.currentChar = parser.source.charCodeAt(++parser.index);
  parser.flags |= Flags.NewLine;
  if ((state & LexerState.LastIsCR) === 0) {
    parser.column = 0;
    parser.line++;
  }
}

export function scanNewLine(parser: Parser): void {
  parser.flags |= Flags.NewLine;
  parser.currentChar = parser.source.charCodeAt(++parser.index);
  parser.column = 0;
  parser.line++;
}

// ECMA-262 11.2 White Space
// gC=Zs, U+0009, U+000B, U+000C, U+FEFF
export function isExoticECMAScriptWhitespace(ch: number): boolean {
  return (
    ch === Chars.NonBreakingSpace ||
    ch === Chars.ZeroWidthNoBreakSpace ||
    ch === Chars.NextLine ||
    ch === Chars.Ogham ||
    (ch >= Chars.EnQuad && ch <= Chars.ZeroWidthSpace) ||
    ch === Chars.NarrowNoBreakSpace ||
    ch === Chars.MathematicalSpace ||
    ch === Chars.IdeographicSpace ||
    ch === Chars.ThinSpace ||
    ch === Chars.ByteOrderMark
  );
}

/**
 * Converts a value to a hex value
 *
 * @param code CodePoint
 */
export function toHex(code: number): number {
  return code < Chars.UpperA ? code - Chars.Zero : (code - Chars.UpperA + 10) & 0xf;
}

/**
 * Converts a token to a string representation
 *
 * @param t Token
 */
export function convertTokenType(t: Token): string {
  switch (t) {
    case Token.NumericLiteral:
      return 'NumericLiteral';
    case Token.StringLiteral:
      return 'StringLiteral';
    case Token.FalseKeyword:
    case Token.TrueKeyword:
      return 'BooleanLiteral';
    case Token.NullKeyword:
      return 'NullLiteral';
    case Token.RegularExpression:
      return 'RegularExpression';
    case Token.TemplateContinuation:
    case Token.TemplateSpan:
    case Token.Template:
      return 'TemplateLiteral';
    default:
      if ((t & Token.IsIdentifier) === Token.IsIdentifier) return 'Identifier';
      if ((t & Token.Keyword) === Token.Keyword) return 'Keyword';

      return 'Punctuator';
  }
}
