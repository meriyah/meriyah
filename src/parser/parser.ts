import { convertTokenType } from '../lexer';
import { Token } from '../token';
import type * as ESTree from '../estree';
import {
  Flags,
  type OnComment,
  type OnInsertedSemicolon,
  type OnToken,
  type Location,
  type AssignmentKind,
  type DestructuringKind,
} from '../common';

export type ParserOptions = {
  shouldAddLoc?: boolean;
  shouldAddRanges?: boolean;
  /**
   * Used together with source maps. File containing the code being parsed
   */
  sourceFile?: string;
  /**
   * Holds either a function or array used on every comment
   */
  onComment?: OnComment;
  /**
   * Function invoked with the character offset when automatic semicolon insertion occurs
   */
  onInsertedSemicolon?: OnInsertedSemicolon;
  /**
   * Holds either a function or array used on every token
   */
  onToken?: OnToken;
};

export class Parser {
  private lastOnToken: [string, number, number, ESTree.SourceLocation] | null = null;

  token = Token.EOF;
  /**
   * The mutable parser flags, in case any flags need passed by reference.
   */
  flags = Flags.None;
  /**
   * The current index
   */
  index = 0;
  /**
   * Beginning of current line
   */
  line = 1;

  /**
   * Beginning of current column
   */
  column = 0;

  /**
   * Start position of whitespace/comment before current token
   */
  startIndex = 0;

  /**
   * The end of the source code
   */
  end = 0;

  /**
   * Start position of text of current token
   */
  tokenIndex = 0;

  /**
   * Start position of the column before newline
   */
  startColumn = 0;

  /**
   * Position in the input code of the first character after the last newline
   */
  tokenColumn = 0;

  /**
   * The number of newlines
   */
  tokenLine = 1;

  /**
   * Start position of text of current token
   */
  startLine = 1;

  /**
   * Holds the scanned token value
   */
  tokenValue: any = '';

  /**
   * Holds the raw text that have been scanned by the lexer
   */
  tokenRaw = '';

  /**
   * Holds the regExp info text that have been collected by the lexer
   */
  tokenRegExp: void | {
    pattern: string;
    flags: string;
  } = void 0;

  /**
   * The code point at the current index
   */
  currentChar = 0;

  /**
   *  https://tc39.es/ecma262/#sec-module-semantics-static-semantics-exportednames
   */
  exportedNames: Record<string, number> = {};

  /**
   * https://tc39.es/ecma262/#sec-exports-static-semantics-exportedbindings
   */

  exportedBindings: Record<string, number> = {};

  /**
   * Assignable state
   */
  assignable: AssignmentKind | DestructuringKind = 1;

  /**
   * Destructuring state
   */
  destructible: AssignmentKind | DestructuringKind = 0;

  /**
   * Holds leading decorators before "export" or "class" keywords
   */
  leadingDecorators: ESTree.Decorator[] = [];

  constructor(
    /**
     * The source code to be parsed
     */
    public readonly source: string,
    public readonly options: ParserOptions = {},
  ) {
    this.end = source.length;
    this.currentChar = source.charCodeAt(0);
  }

  /**
   * Get the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  getToken() {
    return this.token;
  }

  /**
   * Set the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  setToken(value: Token, replaceLast = false) {
    this.token = value;

    const { onToken } = this.options;

    if (onToken) {
      if (value !== Token.EOF) {
        const loc = {
          start: {
            line: this.tokenLine,
            column: this.tokenColumn,
          },
          end: {
            line: this.line,
            column: this.column,
          },
        };

        if (!replaceLast && this.lastOnToken) {
          onToken(...this.lastOnToken);
        }
        this.lastOnToken = [convertTokenType(value), this.tokenIndex, this.index, loc];
      } else {
        if (this.lastOnToken) {
          onToken(...this.lastOnToken);
          this.lastOnToken = null;
        }
      }
    }
    return value;
  }

  get tokenStart(): Location {
    return {
      index: this.tokenIndex,
      line: this.tokenLine,
      column: this.tokenColumn,
    };
  }

  finishNode<T extends ESTree.Node>(node: T, start: Location, end: Location | void): T {
    if (this.options.shouldAddRanges) {
      node.start = start.index;
      const endIndex = end ? end.index : this.startIndex;
      node.end = endIndex;
      node.range = [start.index, endIndex];
    }

    if (this.options.shouldAddLoc) {
      node.loc = {
        start: {
          line: start.line,
          column: start.column,
        },
        end: end ? { line: end.line, column: end.column } : { line: this.startLine, column: this.startColumn },
      };

      if (this.options.sourceFile) {
        node.loc.source = this.options.sourceFile;
      }
    }

    return node;
  }
}

export function pushComment(comments: ESTree.Comment[], options: ParserOptions): OnComment {
  return function (type: ESTree.CommentType, value: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const comment: ESTree.Comment = {
      type,
      value,
    };

    if (options.shouldAddRanges) {
      comment.start = start;
      comment.end = end;
      comment.range = [start, end];
    }
    if (options.shouldAddLoc) {
      comment.loc = loc;
    }
    comments.push(comment);
  };
}

export function pushToken(tokens: any[], options: ParserOptions): OnToken {
  return function (type: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const token: any = {
      token: type,
    };

    if (options.shouldAddRanges) {
      token.start = start;
      token.end = end;
      token.range = [start, end];
    }
    if (options.shouldAddLoc) {
      token.loc = loc;
    }
    tokens.push(token);
  };
}
