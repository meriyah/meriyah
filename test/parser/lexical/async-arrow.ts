import { Context } from '../../../src/common';
import { pass, fail } from '../../test-utils';
import * as t from 'assert';
import { parseSource } from '../../../src/parser';

describe('Lexical - Arrows', () => {
  fail('Lexical - Arrows (fail)', [
    ['(x) => { let x }', Context.None | Context.OptionsLexical],
    ['(x) => { const x = y }', Context.None | Context.OptionsLexical],
    ['([a,b,c]) => { const c = x; }', Context.None | Context.OptionsLexical],
    ['(a, a) => {}', Context.None | Context.OptionsLexical],
    ['(a, b, a) => {}', Context.None | Context.OptionsLexical],
    ['(b, a, a) => {}', Context.None | Context.OptionsLexical],
    ['(a, a, b) => {}', Context.None | Context.OptionsLexical],
    ['(b, a, b, a) => {}', Context.None | Context.OptionsLexical],
    ['(b, a, b, a, [fine]) => {}', Context.None | Context.OptionsLexical],
    ['(b, a, b, a = x) => {}', Context.None | Context.OptionsLexical],
    ['(b, a, b, ...a) => {}', Context.None | Context.OptionsLexical],
    ['([a, a]) => {}', Context.None | Context.OptionsLexical],
    ['([a, b, a]) => {}', Context.None | Context.OptionsLexical],
    ['([b, a, a]) => {}', Context.None | Context.OptionsLexical],
    ['([a, a, b]) => {}', Context.None | Context.OptionsLexical],
    ['([b, a, b, a]) => {}', Context.None | Context.OptionsLexical],
    ['([b, a], b) => {}', Context.None | Context.OptionsLexical],
    ['([b, a], {b}) => {}', Context.None | Context.OptionsLexical],
    ['([b, a], b=x) => {}', Context.None | Context.OptionsLexical],
    ['([a], a, ...b) => {}', Context.None | Context.OptionsLexical],
    ['a(async a => { let a; })', Context.None | Context.OptionsLexical],
    ['a(async (a, a) => { let a; })', Context.None | Context.OptionsLexical],
    ['a(async (a, [a]) => { let a; })', Context.None | Context.OptionsLexical],
    ['async a => { let a; }', Context.None | Context.OptionsLexical],
    ['async => { let async; }', Context.None | Context.OptionsLexical],
    ['async (x) => { let x }', Context.None | Context.OptionsLexical],
    ['async (x) => { const x = y }', Context.None | Context.OptionsLexical],
    ['async ([a,b,c]) => { const c = x; }', Context.None | Context.OptionsLexical],
    ['async (a, a) => {}', Context.None | Context.OptionsLexical],
    ['async (b, a, a) => {}', Context.None | Context.OptionsLexical],
    ['async (b, a, b, a, [fine]) => {}', Context.None | Context.OptionsLexical],
    ['async ([b, a, b, a]) => {}', Context.None | Context.OptionsLexical],
    ['async ([b, a], ...b) => {}', Context.None | Context.OptionsLexical],
    ['async ({"x": x, x: x}) => a', Context.OptionsLexical],
    ['async ({x: y, "x": x = y}) => { let y; }', Context.OptionsLexical],
    ['async ({3: x, 4: x}) => a', Context.OptionsLexical],
    ['async ({x: x, x: x}) => a', Context.OptionsLexical],
    ['async ({x: y, "x": x = y}) => { let y; }', Context.OptionsLexical],
    ['([a,b,c]) => { const c = x; }', Context.None | Context.OptionsLexical],
    ['async b => { let b; }', Context.None | Context.OptionsLexical],
    ['async => { let async; }', Context.None | Context.OptionsLexical],
    ['x = async => { let async; }', Context.None | Context.OptionsLexical],
    ['async (a = b => { let b; })', Context.None | Context.OptionsLexical],
    ['async yield => { let yield; }', Context.None | Context.OptionsLexical],
    ['x = async yield => { let yield; }', Context.None | Context.OptionsLexical],
    ['async await => { let await; }', Context.None | Context.OptionsLexical],
    ['(async => { let async; })', Context.None | Context.OptionsLexical],
    ['(async (a = b => { let b; }))', Context.None | Context.OptionsLexical],
    ['(async yield => { let yield; })', Context.None | Context.OptionsLexical],
    ['(async await => { let await; })', Context.None | Context.OptionsLexical]
  ]);

  for (const arg of [
    '(x) => { function x() {} }',
    '(x) => { var x; }',
    'x => { function x() {} }',
    'x => { var x; }',
    '() => { let foo; }; foo => {}',
    '() => { let foo; }; () => { let foo; }'
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, undefined, Context.None | Context.OptionsLexical);
      });
    });
  }

  for (const arg of [
    '(x) => { function x() {} }',
    '(x) => { var x; }',
    'x => { function x() {} }',
    'x => { var x; }',
    '() => { let foo; }; foo => {}',
    '() => { let foo; }; () => { let foo; }'
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`${arg}`, undefined, Context.OptionsWebCompat | Context.OptionsLexical);
      });
    });
  }
});
