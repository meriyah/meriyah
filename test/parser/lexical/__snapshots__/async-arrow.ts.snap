// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Lexical - Arrows > Lexical - Arrows (fail) > ([a, a, b]) => {} 1`] = `
"SyntaxError [1:6-1:7]: Duplicate binding 'a'
> 1 | ([a, a, b]) => {}
    |       ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > ([a, a]) => {} 1`] = `
"SyntaxError [1:6-1:7]: Duplicate binding 'a'
> 1 | ([a, a]) => {}
    |       ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > ([a, b, a]) => {} 1`] = `
"SyntaxError [1:9-1:10]: Duplicate binding 'a'
> 1 | ([a, b, a]) => {}
    |          ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > ([a,b,c]) => { const c = x; } 1`] = `
"SyntaxError [1:23-1:24]: Duplicate binding 'c'
> 1 | ([a,b,c]) => { const c = x; }
    |                        ^ Duplicate binding 'c'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > ([a,b,c]) => { const c = x; } 2`] = `
"SyntaxError [1:23-1:24]: Duplicate binding 'c'
> 1 | ([a,b,c]) => { const c = x; }
    |                        ^ Duplicate binding 'c'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > ([a], a, ...b) => {} 1`] = `
"SyntaxError [1:6-1:7]: Duplicate binding 'a'
> 1 | ([a], a, ...b) => {}
    |       ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > ([b, a, a]) => {} 1`] = `
"SyntaxError [1:9-1:10]: Duplicate binding 'a'
> 1 | ([b, a, a]) => {}
    |          ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > ([b, a, b, a]) => {} 1`] = `
"SyntaxError [1:12-1:13]: Duplicate binding 'a'
> 1 | ([b, a, b, a]) => {}
    |             ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > ([b, a], {b}) => {} 1`] = `
"SyntaxError [1:11-1:12]: Duplicate binding 'b'
> 1 | ([b, a], {b}) => {}
    |            ^ Duplicate binding 'b'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > ([b, a], b) => {} 1`] = `
"SyntaxError [1:9-1:10]: Duplicate binding 'b'
> 1 | ([b, a], b) => {}
    |          ^ Duplicate binding 'b'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > ([b, a], b=x) => {} 1`] = `
"SyntaxError [1:9-1:10]: Duplicate binding 'b'
> 1 | ([b, a], b=x) => {}
    |          ^ Duplicate binding 'b'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (a, a) => {} 1`] = `
"SyntaxError [1:4-1:5]: Duplicate binding 'a'
> 1 | (a, a) => {}
    |     ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (a, a, b) => {} 1`] = `
"SyntaxError [1:4-1:5]: Duplicate binding 'a'
> 1 | (a, a, b) => {}
    |     ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (a, b, a) => {} 1`] = `
"SyntaxError [1:7-1:8]: Duplicate binding 'a'
> 1 | (a, b, a) => {}
    |        ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (async (a = b => { let b; })) 1`] = `
"SyntaxError [1:24-1:25]: Duplicate binding 'b'
> 1 | (async (a = b => { let b; }))
    |                         ^ Duplicate binding 'b'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (async => { let async; }) 1`] = `
"SyntaxError [1:21-1:22]: Duplicate binding 'async'
> 1 | (async => { let async; })
    |                      ^ Duplicate binding 'async'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (async await => { let await; }) 1`] = `
"SyntaxError [1:7-1:12]: Await expression not allowed in formal parameter
> 1 | (async await => { let await; })
    |        ^^^^^ Await expression not allowed in formal parameter"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (async yield => { let yield; }) 1`] = `
"SyntaxError [1:27-1:28]: Duplicate binding 'yield'
> 1 | (async yield => { let yield; })
    |                            ^ Duplicate binding 'yield'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (b, a, a) => {} 1`] = `
"SyntaxError [1:7-1:8]: Duplicate binding 'a'
> 1 | (b, a, a) => {}
    |        ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (b, a, b, ...a) => {} 1`] = `
"SyntaxError [1:14-1:15]: Duplicate binding 'a'
> 1 | (b, a, b, ...a) => {}
    |               ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (b, a, b, a = x) => {} 1`] = `
"SyntaxError [1:10-1:11]: Duplicate binding 'a'
> 1 | (b, a, b, a = x) => {}
    |           ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (b, a, b, a) => {} 1`] = `
"SyntaxError [1:10-1:11]: Duplicate binding 'a'
> 1 | (b, a, b, a) => {}
    |           ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (b, a, b, a, [fine]) => {} 1`] = `
"SyntaxError [1:10-1:11]: Duplicate binding 'a'
> 1 | (b, a, b, a, [fine]) => {}
    |           ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (x) => { const x = y } 1`] = `
"SyntaxError [1:17-1:18]: Duplicate binding 'x'
> 1 | (x) => { const x = y }
    |                  ^ Duplicate binding 'x'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > (x) => { let x } 1`] = `
"SyntaxError [1:15-1:16]: Duplicate binding 'x'
> 1 | (x) => { let x }
    |                ^ Duplicate binding 'x'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > a(async (a, [a]) => { let a; }) 1`] = `
"SyntaxError [1:14-1:15]: Duplicate binding 'a'
> 1 | a(async (a, [a]) => { let a; })
    |               ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > a(async (a, a) => { let a; }) 1`] = `
"SyntaxError [1:12-1:13]: Duplicate binding 'a'
> 1 | a(async (a, a) => { let a; })
    |             ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > a(async a => { let a; }) 1`] = `
"SyntaxError [1:20-1:21]: Duplicate binding 'a'
> 1 | a(async a => { let a; })
    |                     ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async ([a,b,c]) => { const c = x; } 1`] = `
"SyntaxError [1:29-1:30]: Duplicate binding 'c'
> 1 | async ([a,b,c]) => { const c = x; }
    |                              ^ Duplicate binding 'c'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async ([b, a, b, a]) => {} 1`] = `
"SyntaxError [1:18-1:19]: Duplicate binding 'a'
> 1 | async ([b, a, b, a]) => {}
    |                   ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async ([b, a], ...b) => {} 1`] = `
"SyntaxError [1:19-1:20]: Duplicate binding 'b'
> 1 | async ([b, a], ...b) => {}
    |                    ^ Duplicate binding 'b'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async ({"x": x, x: x}) => a 1`] = `
"SyntaxError [1:20-1:21]: Duplicate binding 'x'
> 1 | async ({"x": x, x: x}) => a
    |                     ^ Duplicate binding 'x'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async ({3: x, 4: x}) => a 1`] = `
"SyntaxError [1:18-1:19]: Duplicate binding 'x'
> 1 | async ({3: x, 4: x}) => a
    |                   ^ Duplicate binding 'x'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async ({x: x, x: x}) => a 1`] = `
"SyntaxError [1:18-1:19]: Duplicate binding 'x'
> 1 | async ({x: x, x: x}) => a
    |                   ^ Duplicate binding 'x'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async ({x: y, "x": x = y}) => { let y; } 1`] = `
"SyntaxError [1:37-1:38]: Duplicate binding 'y'
> 1 | async ({x: y, "x": x = y}) => { let y; }
    |                                      ^ Duplicate binding 'y'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async ({x: y, "x": x = y}) => { let y; } 2`] = `
"SyntaxError [1:37-1:38]: Duplicate binding 'y'
> 1 | async ({x: y, "x": x = y}) => { let y; }
    |                                      ^ Duplicate binding 'y'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async (a = b => { let b; }) 1`] = `
"SyntaxError [1:23-1:24]: Duplicate binding 'b'
> 1 | async (a = b => { let b; })
    |                        ^ Duplicate binding 'b'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async (a, a) => {} 1`] = `
"SyntaxError [1:10-1:11]: Duplicate binding 'a'
> 1 | async (a, a) => {}
    |           ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async (b, a, a) => {} 1`] = `
"SyntaxError [1:13-1:14]: Duplicate binding 'a'
> 1 | async (b, a, a) => {}
    |              ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async (b, a, b, a, [fine]) => {} 1`] = `
"SyntaxError [1:16-1:17]: Duplicate binding 'a'
> 1 | async (b, a, b, a, [fine]) => {}
    |                 ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async (x) => { const x = y } 1`] = `
"SyntaxError [1:23-1:24]: Duplicate binding 'x'
> 1 | async (x) => { const x = y }
    |                        ^ Duplicate binding 'x'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async (x) => { let x } 1`] = `
"SyntaxError [1:21-1:22]: Duplicate binding 'x'
> 1 | async (x) => { let x }
    |                      ^ Duplicate binding 'x'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async (ā,食,食) => { /* 𢭃 */ } 1`] = `
"SyntaxError [1:11-1:12]: Duplicate binding '食'
> 1 | async (ā,食,食) => { /* 𢭃 */ }
    |            ^ Duplicate binding '食'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async => { let async; } 1`] = `
"SyntaxError [1:20-1:21]: Duplicate binding 'async'
> 1 | async => { let async; }
    |                     ^ Duplicate binding 'async'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async => { let async; } 2`] = `
"SyntaxError [1:20-1:21]: Duplicate binding 'async'
> 1 | async => { let async; }
    |                     ^ Duplicate binding 'async'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async a => { let a; } 1`] = `
"SyntaxError [1:18-1:19]: Duplicate binding 'a'
> 1 | async a => { let a; }
    |                   ^ Duplicate binding 'a'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async await => { let await; } 1`] = `
"SyntaxError [1:6-1:11]: Await expression not allowed in formal parameter
> 1 | async await => { let await; }
    |       ^^^^^ Await expression not allowed in formal parameter"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async b => { let b; } 1`] = `
"SyntaxError [1:18-1:19]: Duplicate binding 'b'
> 1 | async b => { let b; }
    |                   ^ Duplicate binding 'b'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > async yield => { let yield; } 1`] = `
"SyntaxError [1:26-1:27]: Duplicate binding 'yield'
> 1 | async yield => { let yield; }
    |                           ^ Duplicate binding 'yield'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > x = async => { let async; } 1`] = `
"SyntaxError [1:24-1:25]: Duplicate binding 'async'
> 1 | x = async => { let async; }
    |                         ^ Duplicate binding 'async'"
`;

exports[`Lexical - Arrows > Lexical - Arrows (fail) > x = async yield => { let yield; } 1`] = `
"SyntaxError [1:30-1:31]: Duplicate binding 'yield'
> 1 | x = async yield => { let yield; }
    |                               ^ Duplicate binding 'yield'"
`;
